django/db/models/sql/where.py

```diff
@@ -35,48 +35,87 @@ class WhereNode(tree.Node):
-        may_need_split = (
+        return (
             (in_negated and self.connector == AND)
             or (not in_negated and self.connector == OR)
             or self.connector == XOR
         )
-        if may_need_split and self.contains_aggregate:
-            return None, self
+
+    def split_having_qualify(self, negated=False, must_group_by=False):
+        """
+        Return three possibly None nodes: one for those parts of self that
+        should be included in the WHERE clause, one for those parts of
+        self that must be included in the HAVING clause, and one for those
+        parts that refer to window functions.
+        """
+        if not self.contains_aggregate and not self.contains_over_clause:
+            return self, None, None
+        in_negated = negated ^ self.negated
+        must_remain_connected = self.must_remain_connected(negated)
+        if (
+            must_remain_connected
+            and self.contains_aggregate
+            and not self.contains_over_clause
+        ):
+            # It's must cheaper to short-circuit and stash everything in the
+            # HAVING clause than split children if possible.
+            return None, self, None
         where_parts = []
         having_parts = []
+        qualify_parts = []
         for c in self.children:
-            if hasattr(c, "split_having"):
-                where_part, having_part = c.split_having(in_negated)
+            if hasattr(c, "split_having_qualify"):
+                where_part, having_part, qualify_part = c.split_having_qualify(
+                    in_negated, must_group_by
+                )
                 if where_part is not None:
                     where_parts.append(where_part)
                 if having_part is not None:
                     having_parts.append(having_part)
+                if qualify_part is not None:
+                    qualify_parts.append(qualify_part)
+            elif c.contains_over_clause:
+                qualify_parts.append(c)
             elif c.contains_aggregate:
                 having_parts.append(c)
             else:
                 where_parts.append(c)
+        if must_remain_connected and qualify_parts:
+            # Disjunctive heterogenous predicates can be pushed down to qualify
+            # as long as no conditional aggregation is involved.
+            if not where_parts or (where_parts and not must_group_by):
+                return None, None, self
+            elif where_parts:
+                # In theory this should only be enforced when dealing with
+                # `where_parts` containing predicates against multi-valued
+                # relationships that could affect aggregation results but
+                # this is complex to infer properly.
+                raise NotImplementedError(
+                    "Heterogenous disjunctive predicates against window functions are "
+                    "not implemented when performing conditional aggregation"
```

Comment:
I understand what you meant, however this message can be difficult for most of folks. Unfortunately, I don't have a better wording (at least for now).

Reply:
Yeah I unfortunately couldn't come with anything better either. The more cases I fixed the more acute the limitations were and the harder to describe.